service/controller/resource/workermigration/create_test.go

```diff
@@ -0,0 +1,236 @@
+package workermigration
+
+import (
+	"context"
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-07-01/compute"
+	"github.com/Azure/go-autorest/autorest"
+	providerv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger/microloggertest"
+	"github.com/giantswarm/to"
+	"github.com/golang/mock/gomock"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	capzv1alpha3 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	expcapzv1alpha3 "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3"
+	capiv1alpha3 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	expcapiv1alpha3 "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/yaml"
+
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+	"github.com/giantswarm/azure-operator/v4/service/controller/resource/workermigration/internal/azure"
+	"github.com/giantswarm/azure-operator/v4/service/controller/resource/workermigration/internal/mock_azure"
+)
+
+//go:generate mockgen -destination internal/mock_azure/api.go -source internal/azure/spec.go API
+
+func TestMigrationCreatesMachinePoolCRs(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	ctrlClient := newFakeClient()
+	m := mock_azure.NewMockAPI(ctrl)
+	r := &Resource{
+		azureapi:   m,
+		ctrlClient: ctrlClient,
+		logger:     microloggertest.New(),
+	}
+
+	ensureCRsExist(t, ctrlClient, []string{
+		"cluster.yaml",
+		"azureconfig.yaml",
+		"azurecluster.yaml",
+	})
+
+	o, err := loadCR("azureconfig.yaml")
+	if err != nil {
+		t.Fatal(err)
+	}
+	cr := o.(*providerv1alpha1.AzureConfig)
+
+	m.
+		EXPECT().
+		GetVMSS(gomock.Any(), key.ResourceGroupName(*cr), key.WorkerVMSSName(*cr)).
+		Return(newBuiltinVMSS(3, key.WorkerVMSSName(*cr)), nil).
+		Times(1)
+
+	m.
+		EXPECT().
+		DeleteVMSS(gomock.Any(), gomock.Any(), gomock.Any()).
+		Times(0)
+
+	err = r.EnsureCreated(context.Background(), cr)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	opts := client.MatchingLabels{
+		capiv1alpha3.ClusterLabelName: key.ClusterName(cr),
+	}
+	mpList := new(expcapiv1alpha3.MachinePoolList)
+	err = ctrlClient.List(context.Background(), mpList, opts)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(mpList.Items) == 0 {
+		t.Fatal("expected at least one MachinePool CR to exist. got 0.")
+	}
+
+	// gomock verifies rest of the assertions on exit.
```

Comment:
This is also slightly nasty in terms of readability or explicitness, but I get the reasons why this is like it is. Probably not a problem once we get used to this?

Reply:
You mean the fact that gomock asserts the number of calls, the arguments passed and so on? I'm used to this behavior from other languages, so I think it's fine, but I reckon is not as explicit as calling a `verify` method.

Reply:
I rather mean the fact that the test fails on mock verification failures only after the whole function executed.